UploadableArtifact withSize()

phasza opened the pull request:
In the RepositoryHandle class one can upload a file with either a given InputStream or a java.io.File
This is rather handy, because one can use the InputStream if he wants to avoid using the java.io library and use the java.nio instead, but also because the inputstream opening-closing is rather messy in this API, and this way the user can control the process.

However, if we try to attach a progress listener to the UploadableArtifact with the withListener(listener) method, an IllegalState exception is thrown asking the user to upload a java.io.File instead of the InputStream because it does not know the content size.

This can be solved rather easily just by setting the content size for the UploadableArtifact before calling the doUpload() method.

This PR adds a new method to the UploadArtifact -> withSize(long size):
Given a listener is attached to the UploadArtifact
- if java.io.File is given, then a progress recorder is created with the file.length() size
- if an InputStream is given and the size is set by the user, then a progress recorder is created with the given size
- if an  InputStream is given and the size was not set by the user, then an IllegalStateException is thrown asking the user to either set the size or to use java.io.File to upload

eyalbe4 commented:
recheckcla

eyalbe4 reviewed (changes requested):
Thanks so much for this contribution!
Can you please sign the CLA as described in the comment added to this PR?
Also, can you please add to the README an example showing how to upload a file, using a progress listener, with and without the new ```withSize``` option?

phasza commented:
I have read the CLA Document and I hereby sign the CLA

phasza commented:
recheckcla

phasza commented:
@eyalbe4 
sorry for taking this long, for such a small change :P
I have rebased the branch and added the description the the README.md

Cheers,
P

ndudekwork commented:
any updates on that?

eyalbe4 commented:
Thanks you for this contribution @ndudekwork and sorry for the time it took to merge it.
